### ISSUE TITLE
Frozen should be type family and not data family

### ISSUE THREAD

**Issue by Shimuuar** [lehins/random#56]
Finally I got time to work on random

`Frozen` shouldn't be a data family since it isn't injective in first place. It's meant to be immutable snapshot of generator state so it shouldn't carry type token in its signature.  For exaple here is mwc-random's analog: `save :: PrimMonad m => Gen (PrimState m) -> m Seed`

With current state of affairs it becomes impossible to return frozen generator from inside of 
`runST` block and reuse in another which defeats purpose of `Frozen`

**Comment by lehins** [comment]
> Frozen shouldn't be a data family since it isn't injective in first place.

Instead of me trying to go on a long debate and try convince you that you are wrong, maybe instead you would like to submit a PR that implements the change you are suggesting? Keep it mind that it is important that ghc is able to unambiguously infer the type of the stateful generator from it's frozen state. With current approach of `data family` I can do:
```haskell
λ> g <- thawGen (PrimGen (mkStdGen 217))
λ> :t g
g :: PrimGen RealWorld StdGen
```
and the compiler is happy to figure out the type of stateful generator for me. With `type family` it can't be done.


> With current state of affairs it becomes impossible to return frozen generator from inside of
runST block and reuse in another which defeats purpose of Frozen

And here again you are wrong. Here is an instance and a helper function that does exactly what you said is impossible:

```haskell
instance (s ~ PrimState m, PrimMonad m) => Random.MonadRandom (Gen s) m where
  newtype Frozen (Gen s) = Frozen Seed
  thawGen (Frozen seed) = restore seed
  freezeGen = fmap Frozen . save
  ... few other functions

withFrozenMWC :: Seed -> (forall s . Gen s -> ST s a) -> (a, Seed)
withFrozenMWC fg action = runST $ do
  g <- Random.thawGen (Frozen fg)
  res <- action g
  Frozen fg' <- Random.freezeGen g
  pure (res, fg')
```

If you are talking about doing the above in general, then yes it is impossible by design. Because with `ST` you get an instance `MonadRandom (Gen s) (ST s)`. Of course, if you have a better design in mind for the whole API and you wanna redo a lot of the work that we did so far, be my guest. Most importantly, if the code you are going to come up with is gonna be more user friendly and performs good, then I'll be the first one to vote to abandon current design.

**Comment by Shimuuar** [comment]
@lehins I created #67 with straightforward fix. 

My point is Frozen is immutable snapshot and if it carry around state tag it becomes mostly useless. At least in conjunction with ST monad. And I think that useful code that require type annotations is better that useless code that doesn't need them. I outlined alternative design, but it's not obvious how it could be implemented without getting into things like `MonadRandom g s m`

> And here again you are wrong. Here is an instance and a helper function that does exactly what you said is impossible:

Not at all! I meant impossible generically. This example is monomorphic in generator type. I could've used save/restore from mwc-random for same effect.
